airflow/decorators/assets.py

```
@@ -0,0 +1,159 @@
+
+    :meta private:
+    """
+
+    function: types.FunctionType
+    schedule: ScheduleArg
+
+    def __attrs_post_init__(self) -> None:
+        parameters = inspect.signature(self.function).parameters
+        if "self" in parameters:
+            self.function: types.FunctionType = _handle_self_argument(cast(types.FunctionType, self.function))  # type: ignore[assignment]
+
+        with DAG(dag_id=self.name, schedule=self.schedule, auto_register=True) as dag:
+            _AssetMainOperator(
+                task_id="__main__",
+                inlets=[
+                    AssetRef(name=inlet_asset_name)
+                    for inlet_asset_name in parameters
+                    if inlet_asset_name not in ("self", "context")
+                ],
+                outlets=[self.to_asset()],
+                python_callable=self.function,
+                definition_name=self.name,
+                uri=self.uri,
+            )
+
+        # TODO: Currently this just gets serialized into a string.
+        # When we create UI for assets, we should add logic to serde so the
+        # serialized DAG contains appropriate asset information.
+        dag._wrapped_definition = self
+
+        DAG.bulk_write_to_db([dag])
+
+    def to_asset(self) -> Asset:
+        return Asset(
+            name=self.name,
+            uri=self.uri,
+            group=self.group,
+            extra=self.extra,
+        )
+
+    def serialize(self):
+        return {
+            "uri": self.uri,
+            "name": self.name,
+            "group": self.group,
+            "extra": self.extra,
+        }
+
+
+@attrs.define(kw_only=True)
+class asset:
+    """Create an asset by decorating a materialization function."""
+
+    schedule: ScheduleArg
+    uri: str | ObjectStoragePath | None = None
+    group: str = attrs.field(
+        kw_only=True,
+        default="",
+        validator=[attrs.validators.max_len(1500), _validate_identifier],
+    )
```

Probably don’t need these anymore since the validation is already done in `Asset`.

---

Hmm actually failing fast here is not a bad idea. We should think of a way to keep the validation in sync here and in Asset though.

---

The only thing I can think of is probably adding a unit test? They already using the same `_validate_identifier` function and `@asset` doesn't have an asset_type by default to fallback to

---

After a second thought, it's probably not as useful for failing fast here. The only thing we do before the check is check the nested function.